apps/voting/contracts/Voting.sol

```diff
@@ -361,43 +513,43 @@ contract Voting is IForwarder, AragonApp {
         vote_.executionScript = _executionScript;
 
         emit StartVote(voteId, msg.sender, _metadata);
-
-        if (_castVote && _canVote(voteId, msg.sender)) {
-            _vote(voteId, true, msg.sender);
-        }
     }
 
     /**
     * @dev Internal function to cast a vote or object to.
       @dev It assumes that voter can support or object to the vote
     */
-    function _vote(uint256 _voteId, bool _supports, address _voter) internal {
+    function _vote(uint256 _voteId, bool _supports, address _voter, bool _isDelegate) internal {
         Vote storage vote_ = votes[_voteId];
 
         // This could re-enter, though we can assume the governance token is not malicious
         uint256 voterStake = token.balanceOfAt(_voter, vote_.snapshotBlock);
         VoterState state = vote_.voters[_voter];
 
         // If voter had previously voted, decrease count
-        if (state == VoterState.Yea) {
+        if (state == VoterState.Yea || state == VoterState.DelegateYea) {
             vote_.yea = vote_.yea.sub(voterStake);
-        } else if (state == VoterState.Nay) {
+        } else if (state == VoterState.Nay || state == VoterState.DelegateNay) {
             vote_.nay = vote_.nay.sub(voterStake);
         }
 
         if (_supports) {
             vote_.yea = vote_.yea.add(voterStake);
-            vote_.voters[_voter] = VoterState.Yea;
+            vote_.voters[_voter] = _isDelegate ? VoterState.DelegateYea : VoterState.Yea;
         } else {
             vote_.nay = vote_.nay.add(voterStake);
-            vote_.voters[_voter] = VoterState.Nay;
+            vote_.voters[_voter] = _isDelegate ? VoterState.DelegateNay : VoterState.Nay;
         }
 
         emit CastVote(_voteId, _voter, _supports, voterStake);
 
         if (_getVotePhase(vote_) == VotePhase.Objection) {
             emit CastObjection(_voteId, _voter, voterStake);
         }
+
+        if (_isDelegate) {
+            emit CastVoteAsDelegate(_voteId, msg.sender, _voter, _supports, voterStake);
```

Comment:
Events in a loop may eat a lot of gas. It's worth considering moving it out of the loop—at least a new one.

Reply:
Moved `CastVoteAsDelegate` up to the `attemptVoteForMultiple` in 32a3dd239c42b642c2cbba3065bbf66e2f4a61b0.